File: backend/runner/tools/task.go
package tools

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/cloudwego/eino/components/tool"
	"github.com/cloudwego/eino/schema"
)

// ========== Task Types ==========

// TaskStatus represents the status of a task
type TaskStatus string

const (
	TaskStatusPending    TaskStatus = "pending"
	TaskStatusInProgress TaskStatus = "in_progress"
	TaskStatusCompleted  TaskStatus = "completed"
	TaskStatusCancelled  TaskStatus = "cancelled"
)

// Task represents a tracked task
type Task struct {
	ID          string     `json:"id"`
	Content     string     `json:"content"`
	Status      TaskStatus `json:"status"`
	Owner       string     `json:"owner,omitempty"`
	BlockedBy   []string   `json:"blocked_by,omitempty"`
	Blocking    []string   `json:"blocking,omitempty"`
	CreatedAt   int64      `json:"created_at"`
	UpdatedAt   int64      `json:"updated_at"`
	CompletedAt *int64     `json:"completed_at,omitempty"`
}

// TaskStore manages tasks in memory and optionally persists to disk
type TaskStore struct {
	mu         sync.RWMutex
	tasks      map[string]*Task
	taskDir    string
}

var (
	globalTaskStore     *TaskStore
	globalTaskStoreOnce sync.Once
)

// GetTaskStore returns the global task store
func GetTaskStore(taskDir ...string) *TaskStore {
	globalTaskStoreOnce.Do(func() {
		dir := "."
		if len(taskDir) > 0 {
			dir = taskDir[0]
		}
		globalTaskStore = &TaskStore{
			tasks:   make(map[string]*Task),
			taskDir: dir,
		}
		globalTaskStore.loadFromDisk()
	})
	return globalTaskStore
}

func (s *TaskStore) loadFromDisk() {
	taskFile := filepath.Join(s.taskDir, ".runner_tasks.json")
	data, err := os.ReadFile(taskFile)
	if err != nil {
		return
	}

	var tasks []*Task
	if err := json.Unmarshal(data, &tasks); err != nil {
		return
	}

	for _, t := range tasks {
		s.tasks[t.ID] = t
	}
}

func (s *TaskStore) saveToDisk() {
	s.mu.RLock()
	defer s.mu.RUnlock()

	tasks := make([]*Task, 0, len(s.tasks))
	for _, t := range s.tasks {
		tasks = append(tasks, t)
	}

	data, err := json.MarshalIndent(tasks, "", "  ")
	if err != nil {
		return
	}

	os.WriteFile(filepath.Join(s.taskDir, ".runner_tasks.json"), data, 0644)
}

func (s *TaskStore) CreateTask(content, owner string, blockedBy []string) (*Task, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	id, _ := generateTaskID()
	now := time.Now().UnixMilli()

	task := &Task{
		ID:        id,
		Content:   content,
		Status:    TaskStatusPending,
		Owner:     owner,
		BlockedBy: blockedBy,
		CreatedAt: now,
		UpdatedAt: now,
	}

	s.tasks[id] = task
	go s.saveToDisk()

	return task, nil
}

func (s *TaskStore) GetTask(id string) *Task {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.tasks[id]
}

func (s *TaskStore) ListTasks() []*Task {
	s.mu.RLock()
	defer s.mu.RUnlock()

	tasks := make([]*Task, 0, len(s.tasks))
	for _, t := range s.tasks {
		tasks = append(tasks, t)
	}
	return tasks
}

func (s *TaskStore) UpdateTask(id string, updates map[string]interface{}) (*Task, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	task, ok := s.tasks[id]
	if !ok {
		return nil, fmt.Errorf("task not found: %s", id)
	}

	now := time.Now().UnixMilli()

	if content, ok := updates["content"].(string); ok {
		task.Content = content
	}
	if status, ok := updates["status"].(string); ok {
		task.Status = TaskStatus(status)
		if task.Status == TaskStatusCompleted {
			task.CompletedAt = &now
		}
	}
	if owner, ok := updates["owner"].(string); ok {
		task.Owner = owner
	}
	if blockedBy, ok := updates["blocked_by"].([]string); ok {
		task.BlockedBy = blockedBy
	}

	task.UpdatedAt = now
	go s.saveToDisk()

	return task, nil
}

func (s *TaskStore) DeleteTask(id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, ok := s.tasks[id]; !ok {
		return fmt.Errorf("task not found: %s", id)
	}

	delete(s.tasks, id)
	go s.saveToDisk()

	return nil
}

func generateTaskID() (string, error) {
	b := make([]byte, 4)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

// ========== TaskCreateTool ==========

// TaskCreateInput for task create tool
type TaskCreateInput struct {
	Content   string   `json:"content"`             // Task description
	Owner     string   `json:"owner,omitempty"`     // Owner of the task
	BlockedBy []string `json:"blocked_by,omitempty"` // Task IDs this is blocked by
}

// TaskCreateTool creates a new task
type TaskCreateTool struct {
	store *TaskStore
}

func NewTaskCreateTool(taskDir ...string) *TaskCreateTool {
	return &TaskCreateTool{store: GetTaskStore(taskDir...)}
}

func (t *TaskCreateTool) Info(ctx context.Context) (*schema.ToolInfo, error) {
	return &schema.ToolInfo{
		Name: "TaskCreate",
		Desc: "Create a new task for tracking progress.",
		ParamsOneOf: schema.NewParamsOneOfByParams(map[string]*schema.ParameterInfo{
			"content": {
				Type:        schema.String,
				Desc:        "Task description",
				Required:    true,
			},
			"owner": {
				Type:        schema.String,
				Desc:        "Owner of the task",
				Required:    false,
			},
			"blocked_by": {
				Type:        schema.Array,
				Desc:        "Task IDs this task is blocked by",
				Required:    false,
			},
		}),
	}, nil
}

func (t *TaskCreateTool) ValidateInput(ctx context.Context, input string) *ValidationResult {
	var taskInput TaskCreateInput
	if err := json.Unmarshal([]byte(input), &taskInput); err != nil {
		return &ValidationResult{Valid: false, Message: fmt.Sprintf("invalid JSON: %v", err), ErrorCode: 1}
	}
	if taskInput.Content == "" {
		return &ValidationResult{Valid: false, Message: "content is required", ErrorCode: 2}
	}
	return &ValidationResult{Valid: true}
}

func (t *TaskCreateTool) InvokableRun(ctx context.Context, input string, opts ...tool.Option) (string, error) {
	var taskInput TaskCreateInput
	if err := json.Unmarshal([]byte(input), &taskInput); err != nil {
		return "", fmt.Errorf("invalid input: %w", err)
	}

	task, err := t.store.CreateTask(taskInput.Content, taskInput.Owner, taskInput.BlockedBy)
	if err != nil {
		return "", err
	}

	result, _ := json.Marshal(task)
	return string(result), nil
}

// ========== TaskGetTool ==========

// TaskGetInput for task get tool
type TaskGetInput struct {
	ID string `json:"id"` // Task ID
}

// TaskGetTool retrieves a task by ID
type TaskGetTool struct {
	store *TaskStore
}

func NewTaskGetTool(taskDir ...string) *TaskGetTool {
	return &TaskGetTool{store: GetTaskStore(taskDir...)}
}

func (t *TaskGetTool) Info(ctx context.Context) (*schema.ToolInfo, error) {
	return &schema.ToolInfo{
		Name: "TaskGet",
		Desc: "Retrieve a task by its ID with full details.",
		ParamsOneOf: schema.NewParamsOneOfByParams(map[string]*schema.ParameterInfo{
			"id": {
				Type:        schema.String,
				Desc:        "Task ID to retrieve",
				Required:    true,
			},
		}),
	}, nil
}

func (t *TaskGetTool) ValidateInput(ctx context.Context, input string) *ValidationResult {
	var taskInput TaskGetInput
	if err := json.Unmarshal([]byte(input), &taskInput); err != nil {
		return &ValidationResult{Valid: false, Message: fmt.Sprintf("invalid JSON: %v", err), ErrorCode: 1}
	}
	if taskInput.ID == "" {
		return &ValidationResult{Valid: false, Message: "id is required", ErrorCode: 2}
	}
	return &ValidationResult{Valid: true}
}

func (t *TaskGetTool) InvokableRun(ctx context.Context, input string, opts ...tool.Option) (string, error) {
	var taskInput TaskGetInput
	if err := json.Unmarshal([]byte(input), &taskInput); err != nil {
		return "", fmt.Errorf("invalid input: %w", err)
	}

	task := t.store.GetTask(taskInput.ID)
	if task == nil {
		return "", fmt.Errorf("task not found: %s", taskInput.ID)
	}

	result, _ := json.Marshal(task)
	return string(result), nil
}

// ========== TaskListTool ==========

// TaskListInput for task list tool
type TaskListInput struct {
	Status string `json:"status,omitempty"` // Filter by status
	Owner  string `json:"owner,omitempty"`  // Filter by owner
}

// TaskListTool lists all tasks
type TaskListTool struct {
	store *TaskStore
}

func NewTaskListTool(taskDir ...string) *TaskListTool {
	return &TaskListTool{store: GetTaskStore(taskDir...)}
}

func (t *TaskListTool) Info(ctx context.Context) (*schema.ToolInfo, error) {
	return &schema.ToolInfo{
		Name: "TaskList",
		Desc: "List all tasks in the task list.",
		ParamsOneOf: schema.NewParamsOneOfByParams(map[string]*schema.ParameterInfo{
			"status": {
				Type:        schema.String,
				Desc:        "Filter by status (pending, in_progress, completed, cancelled)",
				Required:    false,
			},
			"owner": {
				Type:        schema.String,
				Desc:        "Filter by owner",
				Required:    false,
			},
		}),
	}, nil
}

func (t *TaskListTool) ValidateInput(ctx context.Context, input string) *ValidationResult {
	return &ValidationResult{Valid: true}
}

func (t *TaskListTool) InvokableRun(ctx context.Context, input string, opts ...tool.Option) (string, error) {
	var taskInput TaskListInput
	if input != "" {
		json.Unmarshal([]byte(input), &taskInput)
	}

	tasks := t.store.ListTasks()

	// Filter
	var filtered []*Task
	for _, task := range tasks {
		if taskInput.Status != "" && string(task.Status) != taskInput.Status {
			continue
		}
		if taskInput.Owner != "" && task.Owner != taskInput.Owner {
			continue
		}
		filtered = append(filtered, task)
	}

	type ListOutput struct {
		Tasks []*Task `json:"tasks"`
		Count int     `json:"count"`
	}

	output := ListOutput{Tasks: filtered, Count: len(filtered)}
	result, _ := json.Marshal(output)
	return string(result), nil
}

// ========== TaskUpdateTool ==========

// TaskUpdateInput for task update tool
type TaskUpdateInput struct {
	ID      string `json:"id"`                // Task ID
	Status  string `json:"status,omitempty"` // New status
	Owner   string `json:"owner,omitempty"`  // New owner
	Content string `json:"content,omitempty"` // New content
}

// TaskUpdateTool updates a task
type TaskUpdateTool struct {
	store *TaskStore
}

func NewTaskUpdateTool(taskDir ...string) *TaskUpdateTool {
	return &TaskUpdateTool{store: GetTaskStore(taskDir...)}
}

func (t *TaskUpdateTool) Info(ctx context.Context) (*schema.ToolInfo, error) {
	return &schema.ToolInfo{
		Name: "TaskUpdate",
		Desc: "Update a task's status, details, or ownership.",
		ParamsOneOf: schema.NewParamsOneOfByParams(map[string]*schema.ParameterInfo{
			"id": {
				Type:        schema.String,
				Desc:        "Task ID to update",
				Required:    true,
			},
			"status": {
				Type:        schema.String,
				Desc:        "New status (pending, in_progress, completed, cancelled)",
				Required:    false,
			},
			"owner": {
				Type:        schema.String,
				Desc:        "New owner",
				Required:    false,
			},
			"content": {
				Type:        schema.String,
				Desc:        "New task content",
				Required:    false,
			},
		}),
	}, nil
}

func (t *TaskUpdateTool) ValidateInput(ctx context.Context, input string) *ValidationResult {
	var taskInput TaskUpdateInput
	if err := json.Unmarshal([]byte(input), &taskInput); err != nil {
		return &ValidationResult{Valid: false, Message: fmt.Sprintf("invalid JSON: %v", err), ErrorCode: 1}
	}
	if taskInput.ID == "" {
		return &ValidationResult{Valid: false, Message: "id is required", ErrorCode: 2}
	}
	return &ValidationResult{Valid: true}
}

func (t *TaskUpdateTool) InvokableRun(ctx context.Context, input string, opts ...tool.Option) (string, error) {
	var taskInput TaskUpdateInput
	if err := json.Unmarshal([]byte(input), &taskInput); err != nil {
		return "", fmt.Errorf("invalid input: %w", err)
	}

	updates := make(map[string]interface{})
	if taskInput.Status != "" {
		updates["status"] = taskInput.Status
	}
	if taskInput.Owner != "" {
		updates["owner"] = taskInput.Owner
	}
	if taskInput.Content != "" {
		updates["content"] = taskInput.Content
	}

	task, err := t.store.UpdateTask(taskInput.ID, updates)
	if err != nil {
		return "", err
	}

	result, _ := json.Marshal(task)
	return string(result), nil
}

// ========== TodoWriteTool ==========

// TodoInput for todo write tool
type TodoInput struct {
	Action string            `json:"action"`             // Action: "add", "done", "remove", "clear"
	Content string           `json:"content,omitempty"`  // Todo content (for add)
	Index  int               `json:"index,omitempty"`   // Index (for done/remove)
	Todos  []map[string]string `json:"todos,omitempty"`  // Full todo list (for clear)
}

// TodoItem represents a single todo item
type TodoItem struct {
	Content string `json:"content"`
	Done   bool   `json:"done"`
}

// TodoWriteTool manages a simple todo list
type TodoWriteTool struct {
	store   *TaskStore
	todos   []TodoItem
	todosMu sync.RWMutex
}

func NewTodoWriteTool(taskDir ...string) *TodoWriteTool {
	return &TodoWriteTool{
		store: GetTaskStore(taskDir...),
		todos: make([]TodoItem, 0),
	}
}

func init() {
	// Task tools - Read/Write mix
	GlobalRegistry.Register(ToolMeta{
		Name:           "TaskCreate",
		Desc:           "Create a new task for tracking progress.",
		IsReadOnly:     false,
		MaxResultChars: 500,
		DefaultRisk:    "low",
		Creator: func(basePath string) interface{} {
			return NewTaskCreateTool(basePath)
		},
	})
	GlobalRegistry.Register(ToolMeta{
		Name:           "TaskGet",
		Desc:           "Retrieve a task by its ID with full details.",
		IsReadOnly:     true,
		MaxResultChars: 1000,
		DefaultRisk:    "low",
		Creator: func(basePath string) interface{} {
			return NewTaskGetTool(basePath)
		},
	})
	GlobalRegistry.Register(ToolMeta{
		Name:           "TaskList",
		Desc:           "List all tasks in the task list.",
		IsReadOnly:     true,
		MaxResultChars: 5000,
		DefaultRisk:    "low",
		Creator: func(basePath string) interface{} {
			return NewTaskListTool(basePath)
		},
	})
	GlobalRegistry.Register(ToolMeta{
		Name:           "TaskUpdate",
		Desc:           "Update a task's status, details, or ownership.",
		IsReadOnly:     false,
		MaxResultChars: 500,
		DefaultRisk:    "low",
		Creator: func(basePath string) interface{} {
			return NewTaskUpdateTool(basePath)
		},
	})
	GlobalRegistry.Register(ToolMeta{
		Name:           "TodoWrite",
		Desc:           "Create and manage a todo list for tracking progress.",
		IsReadOnly:     false,
		MaxResultChars: 2000,
		DefaultRisk:    "low",
		Creator: func(basePath string) interface{} {
			return NewTodoWriteTool(basePath)
		},
	})
}

func (t *TodoWriteTool) Info(ctx context.Context) (*schema.ToolInfo, error) {
	return &schema.ToolInfo{
		Name: "TodoWrite",
		Desc: "Create and manage a todo list for tracking progress.",
		ParamsOneOf: schema.NewParamsOneOfByParams(map[string]*schema.ParameterInfo{
			"action": {
				Type:        schema.String,
				Desc:        "Action: add (add todo), done (mark complete), remove (delete), clear (replace all)",
				Required:    true,
			},
			"content": {
				Type:        schema.String,
				Desc:        "Todo content (for add action)",
				Required:    false,
			},
			"index": {
				Type:        schema.Integer,
				Desc:        "Todo index (for done/remove actions, 0-indexed)",
				Required:    false,
			},
			"todos": {
				Type:        schema.Array,
				Desc:        "Full todo list (for clear action)",
				Required:    false,
			},
		}),
	}, nil
}

func (t *TodoWriteTool) ValidateInput(ctx context.Context, input string) *ValidationResult {
	var todoInput TodoInput
	if err := json.Unmarshal([]byte(input), &todoInput); err != nil {
		return &ValidationResult{Valid: false, Message: fmt.Sprintf("invalid JSON: %v", err), ErrorCode: 1}
	}
	if todoInput.Action == "" {
		return &ValidationResult{Valid: false, Message: "action is required", ErrorCode: 2}
	}
	validActions := map[string]bool{"add": true, "done": true, "remove": true, "clear": true}
	if !validActions[todoInput.Action] {
		return &ValidationResult{Valid: false, Message: "action must be one of: add, done, remove, clear", ErrorCode: 3}
	}
	return &ValidationResult{Valid: true}
}

func (t *TodoWriteTool) InvokableRun(ctx context.Context, input string, opts ...tool.Option) (string, error) {
	var todoInput TodoInput
	if err := json.Unmarshal([]byte(input), &todoInput); err != nil {
		return "", fmt.Errorf("invalid input: %w", err)
	}

	t.todosMu.Lock()
	defer t.todosMu.Unlock()

	switch todoInput.Action {
	case "add":
		t.todos = append(t.todos, TodoItem{Content: todoInput.Content, Done: false})

	case "done":
		if todoInput.Index >= 0 && todoInput.Index < len(t.todos) {
			t.todos[todoInput.Index].Done = true
		}

	case "remove":
		if todoInput.Index >= 0 && todoInput.Index < len(t.todos) {
			t.todos = append(t.todos[:todoInput.Index], t.todos[todoInput.Index+1:]...)
		}

	case "clear":
		t.todos = make([]TodoItem, 0)
		for _, todo := range todoInput.Todos {
			t.todos = append(t.todos, TodoItem{
				Content: todo["content"],
				Done:    todo["done"] == "true" || todo["done"] == "1",
			})
		}
	}

	type TodoOutput struct {
		Todos []TodoItem `json:"todos"`
		Count int        `json:"count"`
	}

	output := TodoOutput{Todos: t.todos, Count: len(t.todos)}
	result, _ := json.Marshal(output)
	return string(result), nil
}

File: backend/runner-langchain/app/tools/task.py
"""Task tools for task tracking.

Based on Go runner's task.go design.
Provides TaskCreate, TaskGet, TaskList, TaskUpdate tools.
"""

import json
import os
import secrets
import threading
import time
from dataclasses import dataclass, asdict
from enum import Enum
from pathlib import Path
from typing import Optional

from app.tools.registry import get_registry
from app.utils.dir import get_base_dir

logger = __name__

TASK_FILE = ".runner_tasks.json"


class TaskStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


@dataclass
class Task:
    id: str
    content: str
    status: TaskStatus
    owner: Optional[str] = None
    blocked_by: list[str] = None
    created_at: int = 0
    updated_at: int = 0
    completed_at: Optional[int] = None

    def __post_init__(self):
        if self.blocked_by is None:
            self.blocked_by = []


class TaskStore:
    """Thread-safe task store with disk persistence."""

    _instance: Optional["TaskStore"] = None
    _lock = threading.Lock()

    def __init__(self, task_dir: str = "."):
        self.task_dir = Path(task_dir)
        self.tasks: dict[str, Task] = {}
        self._lock = threading.Lock()
        self._load_from_disk()

    @classmethod
    def get_instance(cls, task_dir: Optional[str] = None) -> "TaskStore":
        """Get singleton task store instance."""
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    if task_dir is None:
                        task_dir = str(get_base_dir())
                    cls._instance = cls(task_dir)
        return cls._instance

    def _task_file(self) -> Path:
        return self.task_dir / TASK_FILE

    def _load_from_disk(self) -> None:
        """Load tasks from disk."""
        task_file = self._task_file()
        if not task_file.exists():
            return

        try:
            data = json.loads(task_file.read_text())
            for t_data in data:
                t_data["status"] = TaskStatus(t_data["status"])
                task = Task(**t_data)
                self.tasks[task.id] = task
        except (json.JSONDecodeError, KeyError, ValueError):
            pass

    def _save_to_disk(self) -> None:
        """Save tasks to disk."""
        task_file = self._task_file()
        tasks_data = [asdict(t) for t in self.tasks.values()]
        # Convert enums to strings for JSON
        for t_data in tasks_data:
            t_data["status"] = t_data["status"].value
        try:
            task_file.write_text(json.dumps(tasks_data, indent=2))
        except OSError as e:
            logger.warning("Failed to save tasks: %s", e)

    def _generate_id(self) -> str:
        """Generate a unique task ID."""
        return secrets.token_hex(4)

    def create_task(self, content: str, owner: str = "", blocked_by: list[str] = None) -> Task:
        """Create a new task."""
        with self._lock:
            now = int(time.time() * 1000)
            task = Task(
                id=self._generate_id(),
                content=content,
                status=TaskStatus.PENDING,
                owner=owner,
                blocked_by=blocked_by or [],
                created_at=now,
                updated_at=now,
            )
            self.tasks[task.id] = task
            self._save_to_disk()
            return task

    def get_task(self, task_id: str) -> Optional[Task]:
        """Get a task by ID."""
        with self._lock:
            return self.tasks.get(task_id)

    def list_tasks(self, status: Optional[str] = None, owner: Optional[str] = None) -> list[Task]:
        """List tasks with optional filters."""
        with self._lock:
            tasks = list(self.tasks.values())
            if status:
                tasks = [t for t in tasks if t.status.value == status]
            if owner:
                tasks = [t for t in tasks if t.owner == owner]
            return tasks

    def update_task(self, task_id: str, updates: dict) -> Optional[Task]:
        """Update a task."""
        with self._lock:
            task = self.tasks.get(task_id)
            if not task:
                return None

            now = int(time.time() * 1000)
            if "content" in updates:
                task.content = updates["content"]
            if "status" in updates:
                task.status = TaskStatus(updates["status"])
                if task.status == TaskStatus.COMPLETED:
                    task.completed_at = now
            if "owner" in updates:
                task.owner = updates["owner"]
            if "blocked_by" in updates:
                task.blocked_by = updates["blocked_by"]

            task.updated_at = now
            self._save_to_disk()
            return task

    def delete_task(self, task_id: str) -> bool:
        """Delete a task."""
        with self._lock:
            if task_id in self.tasks:
                del self.tasks[task_id]
                self._save_to_disk()
                return True
            return False


# Global store
_task_store: Optional[TaskStore] = None


def get_task_store() -> TaskStore:
    """Get the global task store."""
    global _task_store
    if _task_store is None:
        _task_store = TaskStore.get_instance()
    return _task_store


# ========== Tool Handlers ==========


def _task_create(content: str, owner: str = "", blocked_by: list[str] = None) -> str:
    """Create a new task."""
    try:
        store = get_task_store()
        task = store.create_task(content, owner, blocked_by)
        return json.dumps({
            "id": task.id,
            "content": task.content,
            "status": task.status.value,
            "owner": task.owner,
            "blocked_by": task.blocked_by,
            "created_at": task.created_at,
            "updated_at": task.updated_at,
        })
    except Exception as e:
        return json.dumps({"error": str(e)})


def _task_get(task_id: str) -> str:
    """Get a task by ID."""
    try:
        store = get_task_store()
        task = store.get_task(task_id)
        if not task:
            return json.dumps({"error": f"Task not found: {task_id}"})

        return json.dumps({
            "id": task.id,
            "content": task.content,
            "status": task.status.value,
            "owner": task.owner,
            "blocked_by": task.blocked_by,
            "created_at": task.created_at,
            "updated_at": task.updated_at,
            "completed_at": task.completed_at,
        })
    except Exception as e:
        return json.dumps({"error": str(e)})


def _task_list(status: str = "", owner: str = "") -> str:
    """List all tasks."""
    try:
        store = get_task_store()
        tasks = store.list_tasks(status or None, owner or None)

        return json.dumps({
            "tasks": [
                {
                    "id": t.id,
                    "content": t.content,
                    "status": t.status.value,
                    "owner": t.owner,
                    "blocked_by": t.blocked_by,
                    "created_at": t.created_at,
                    "updated_at": t.updated_at,
                }
                for t in tasks
            ],
            "count": len(tasks),
        })
    except Exception as e:
        return json.dumps({"error": str(e)})


def _task_update(task_id: str, status: str = "", owner: str = "", content: str = "") -> str:
    """Update a task."""
    try:
        store = get_task_store()
        updates = {}
        if status:
            updates["status"] = status
        if owner:
            updates["owner"] = owner
        if content:
            updates["content"] = content

        task = store.update_task(task_id, updates)
        if not task:
            return json.dumps({"error": f"Task not found: {task_id}"})

        return json.dumps({
            "id": task.id,
            "content": task.content,
            "status": task.status.value,
            "owner": task.owner,
            "blocked_by": task.blocked_by,
            "created_at": task.created_at,
            "updated_at": task.updated_at,
        })
    except Exception as e:
        return json.dumps({"error": str(e)})


def register_tools() -> None:
    """Register task tools."""
    registry = get_registry()

    # TaskCreate
    registry.register(
        name="task_create",
        description="Create a new task for tracking progress.",
        schema={
            "type": "object",
            "properties": {
                "content": {
                    "type": "string",
                    "description": "Task description",
                },
                "owner": {
                    "type": "string",
                    "description": "Owner of the task",
                },
                "blocked_by": {
                    "type": "array",
                    "items": {"type": "string"},
                    "description": "Task IDs this task is blocked by",
                },
            },
            "required": ["content"],
        },
        handler=lambda **kwargs: _task_create(
            content=kwargs.get("content", ""),
            owner=kwargs.get("owner", ""),
            blocked_by=kwargs.get("blocked_by"),
        ),
    )

    # TaskGet
    registry.register(
        name="task_get",
        description="Retrieve a task by its ID with full details.",
        schema={
            "type": "object",
            "properties": {
                "task_id": {
                    "type": "string",
                    "description": "Task ID to retrieve",
                },
            },
            "required": ["task_id"],
        },
        handler=lambda **kwargs: _task_get(kwargs.get("task_id", "")),
    )

    # TaskList
    registry.register(
        name="task_list",
        description="List all tasks in the task list.",
        schema={
            "type": "object",
            "properties": {
                "status": {
                    "type": "string",
                    "description": "Filter by status (pending, in_progress, completed, cancelled)",
                },
                "owner": {
                    "type": "string",
                    "description": "Filter by owner",
                },
            },
            "required": [],
        },
        handler=lambda **kwargs: _task_list(
            status=kwargs.get("status", ""),
            owner=kwargs.get("owner", ""),
        ),
    )

    # TaskUpdate
    registry.register(
        name="task_update",
        description="Update a task's status, details, or ownership.",
        schema={
            "type": "object",
            "properties": {
                "task_id": {
                    "type": "string",
                    "description": "Task ID to update",
                },
                "status": {
                    "type": "string",
                    "description": "New status (pending, in_progress, completed, cancelled)",
                },
                "owner": {
                    "type": "string",
                    "description": "New owner",
                },
                "content": {
                    "type": "string",
                    "description": "New task content",
                },
            },
            "required": ["task_id"],
        },
        handler=lambda **kwargs: _task_update(
            task_id=kwargs.get("task_id", ""),
            status=kwargs.get("status", ""),
            owner=kwargs.get("owner", ""),
            content=kwargs.get("content", ""),
        ),
    )